Puck issue 1456 overlay position

shannonhochkins opened the pull request:
Closes [#1456](https://github.com/puckeditor/puck/issues/1456)

## Description

Fix overlay positioning to fully cover components even when parents/targets apply CSS transforms or fixed positioning. We now rely on the browser’s transformed bounding boxes, add fixed-target handling, and resync overlays when style/class or geometry changes. Removed the legacy inverse-scale correction because getBoundingClientRect already reports transformed sizes/positions, so dividing by accumulated transforms caused misalignment under parent transforms.


## Changes made

- Use the element’s transformed getBoundingClientRect plus scroll offsets to set overlay left/top/width/height; drop accumulateTransform inverse-scaling
  - getBoundingClientRect() already returns the post-transform box: position, width, and height are in the final rendered coordinate space, including all parent transforms (scale/translate/rotate). Dividing by an accumulated scale double-corrects and shrinks/shifts the overlay when parents are transformed.
- Detect fixed ancestors and set overlay position: fixed while skipping scroll offsets in that case.
- Add a MutationObserver watching style and class to resync overlays when styles change.
- Add a lightweight rAF rect watcher to resync when geometry moves due to transforms (e.g., carousel track translate/scale) without waiting for scroll/resize.
- Keep ResizeObserver and ref reattachment logic; sync now depends only on getStyle.
- Supports css transforms completely (scale, rotation, translation, even skew) (example screenshot below, try the same css in production ;) )

<img width="1502" height="541" alt="image" src="https://github.com/user-attachments/assets/53a701b0-a3cb-4d11-824c-dc8bfc47b2b4" />


## How to test

- Update inline css of any component, change to position fixed, overlay should be positioned correctly
- Update inline css of any component, or even a parent, add css transform `scale3d(0.79, 0.79, 1) translate3d(20%, 0, 0);` and you should see the overlay move to match the new position when hovering, even other sub components if added to a parent with transforms like scale3d(0.79, 0.79, 1) translate3d(20%, 0, 0); overlay should match the rendered element bounds.
- Set a draggable ancestor to position: fixed; overlay should also be fixed and align while scrolling.
- Apply style/class changes (e.g., add/remove CSS that shifts size/position) and confirm the overlay updates immediately.
- transitions that manipulate inline styles even not on the target will now respect the new position and overlays will update (carousel example)
- Test disableIframe=true to ensure all the above still works when no iframe is provided

I have done EXTENSIVE testing on this and it's pretty solid! Can't find any more issues

- Multiple tests were done, comparing production to this branch, screen recordings below: 

https://github.com/user-attachments/assets/590a3312-990c-440c-97f4-382a2205ec16

^ Fixes a bug in production in the puck demo that's present

And here's the difference in my software:

- EDIT - This screen recording was taken before i added the raf helper, you can see the overlay sometimes jumps when the carousel moves, until you move your cursor or scroll/resize - this is fixed now, additionally the "fixed" popup overlay that animates, is also patched and now is pinned to the element directly (this was a separate side effect i removed with a way over-engineered solution) - now much simpler and very little changes to puck

https://github.com/user-attachments/assets/2d4699ab-9c19-4290-b7cf-e63bb2cc2f3d

chrisvxd commented on packages/core/components/DraggableComponent/index.tsx:
```diff
@@ -311,7 +324,61 @@ export const DraggableComponent = ({
+          ["style", "class"].includes(mutation.attributeName || "")
+        )
+      ) {
+        sync();
```
We should probably put this inside a `requestAnimationFrame`

chrisvxd reviewed (changes requested):
Sorry for slow review. Can confirm it works, but have performance concerns with the animation loop. See comment.

shannonhochkins commented:
@chrisvxd I've made a few changes, are you able to check them again for me?

**Perf change:** removed the always-on overlay sync loop and switched to event-driven + throttled measurement.

Previously, each DraggableComponent started a requestAnimationFrame loop that ran continuously, even when the app was otherwise idle. That meant we were calling getBoundingClientRect() every frame and occasionally updating overlay state, which kept the renderer active and raised baseline CPU usage.

In this update:

- Overlay sync is now coalesced via scheduleSync() (single rAF per frame). This prevents multiple rapid triggers from causing multiple React updates.

- Resize handling stays event-driven via ResizeObserver, which calls scheduleSync() when the element’s dimensions change.

- Scroll / window resize are handled via listeners (scroll uses capture to catch nested scroll containers), triggering scheduleSync() only when those events occur.

- To still support position changes caused by animations/layout shifts/transforms (i.e. movement that doesn’t emit a scroll/resize/resizeObserver event), we added a throttled rect sampling loop while the component is active (measuring at ~10fps via MEASURE_EVERY_MS). When a rect delta is detected, we trigger scheduleSync().

**Net effect:** we keep overlay positioning correct during real movement (scroll/resize/animations), but avoid the previous “always hot” 60fps loop that increased idle CPU.

The throttled measurement loop is not started on hover, only when selected/dragging, to avoid many components spinning up trackers as the mouse moves through the editor.
